### PR TITLE
Implement, test, and benchmark new proposals

### PR DESCRIPTION
**Pull request by conectado**
# This PR addresses #1748 and #1864

In this PR I did the following things:
* For existing proposals updated the names to closely match those in the handbook
  * Text -> Signal
  * Spending -> Funding Request
  * Set Validator Count -> Set Max Validator Count
  * Add Working Group Leader Opening -> Create Working Group Lead Opening
  * Fill Working Group Leader Opening -> Fill Working Group Lead Opening
  * Set Working Group Budget Capacity ->  Update Working Group Budget
  * Decrease Working Group Leader Stake -> Decrease Working Group Lead Stake
  * Slash Working Group Leader Stake -> Slash Working Group Lead
  * Set Working Group Leader Reward -> Set Working Group Lead Reward
  * Terminate Working Group Leader Role -> Terminate Working Group Lead
* Removed old proposals no longer in the handbook
* Added new proposals:
  * Cancel Working Group Lead Opening
  * Set Membership Price
  * Set Council Budget Increment
  * Set Councilor Reward
  * Set Initial Invitation Balance
  * Set Initial Invitation Count
  * Set Membership Lead Invitation Quota
  * Set Referral Cut
  * **Note:** _Update Budget Request_ and _Funding Request_ changed significantly
* Needed to change the `Balance` type in `Council` to the `Balance` in the `balances` pallet to be able to update the budget of `WorkingGroup` with the budget from `Council`
* Updated `councilor_reward` and `budget_incrment` to be part of the storage as to be able to be updated with `Set Council Budget Increment` and `Set Councilor Reward`

### Also, I have a couple of open questions:
* In [`Set Max Validator Count`](https://joystream.gitbook.io/joystream-handbook/governance/proposals#set-max-validator-count) we use `<staking::Module<T>>::minimum_validator_count()` instead of `MIN_VALIDATOR_COUNT`. Is this okay?

### [Handbook companion PR](https://github.com/Joystream/handbook/pull/27)

### CODE REVIEW

File: runtime-modules/proposals/codex/src/lib.rs [hunk 1]
```diff
@@ -58,40 +58,92 @@ mod types;
...
+
 /// 'Proposals codex' substrate module Trait
 pub trait Trait:
     frame_system::Trait
     + proposals_engine::Trait
     + proposals_discussion::Trait
     + common::Trait
+    + council::Trait
     + staking::Trait
+    + working_group::Trait<ForumWorkingGroupInstance>
```
Comment (shamil-gadelshin) [inline runtime-modules/proposals/codex/src/lib.rs]: This will require the codex update for each new working group created.

File: runtime/src/proposals_configuration/sample_proposal_parameters.json [hunk 2]
```diff
@@ -1,24 +1,25 @@
...
-  "fill_working_group_opening_proposal": {},
-  "set_working_group_budget_capacity_proposal": {},
-  "decrease_working_group_leader_stake_proposal": {},
-  "slash_working_group_leader_stake_proposal": {},
-  "set_working_group_leader_reward_proposal": {},
-  "terminate_working_group_leader_role_proposal": {},
-  "amend_constitution_proposal": {}
-}
\ No newline at end of file
+    "_comment": "This is a sample 'Proposal parameters' JSON file to enable conditional runtime compilation.",
```
Comment (shamil-gadelshin) [inline runtime/src/proposals_configuration/sample_proposal_parameters.json:2]: Please, modify this file to reflect your changes - it is needed by the integration tests.

Comment (shamil-gadelshin) [reply runtime/src/proposals_configuration/sample_proposal_parameters.json:2]: Make sure that this file works, you can use this `proposal_parameters_are_initialized` test for each proposal locally (but don't create tests for each proposal - we won't enable it by default because of the conditional compilation).

File: runtime-modules/working-group/src/lib.rs [hunk 3]
```diff
@@ -63,6 +63,18 @@ use staking_handler::StakingHandler;
 
 type WeightInfoWorkingGroup<T, I> = <T as Trait<I>>::WeightInfo;
 
+/// The forum working group instance alias.
```
Comment (shamil-gadelshin) [inline runtime-modules/working-group/src/lib.rs]: You made the working group - instance-dependent which downgrades its flexibility. For each new working group - you should update it.

File: runtime-modules/proposals/codex/src/types.rs [hunk 4]
```diff
@@ -96,43 +136,43 @@ pub struct GeneralProposalParams<MemberId, AccountId, BlockNumber> {
 #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
 #[derive(Encode, Decode, Clone, PartialEq, Debug)]
 pub struct TerminateRoleParameters<WorkerId, Balance> {
-    /// Leader worker id to fire.
+    /// Worker identifier.
     pub worker_id: WorkerId,
 
-    /// Terminate role slash penalty.
-    pub penalty: Option<Penalty<Balance>>,
+    /// Optional amount to be slashed.
```
Comment (shamil-gadelshin) [inline runtime-modules/proposals/codex/src/types.rs:142]: Why did you remove the penalty here? My guess because the gitbook doesn't contain it. But termination slash is similar to the regular slash. And it contains the slashing rationale. You could confirm it with Bedeho.

File: runtime-modules/proposals/codex/src/benchmarking.rs [hunk 5]
```diff
@@ -0,0 +1,811 @@
...
+        let (account_id, member_id, general_proposal_paramters) = create_proposal_parameters::<T>(t, d);
+
+        let proposal_details = ProposalDetails::SetReferralCut(One::one());
+    }: create_proposal(RawOrigin::Signed(account_id.clone()), general_proposal_paramters, proposal_details.clone())
+    verify {
+        create_proposal_verify::<T>(account_id, member_id, proposal_details);
+    }
+
+    update_working_group_budget_positive_forum {
+        set_wg_and_council_budget::<T>(100, WorkingGroup::Forum);
```
Comment (shamil-gadelshin) [inline runtime-modules/proposals/codex/src/benchmarking.rs:506]: Is it possible to introduce some parameter x=0..1 which could define the balance sign? This could reduce the benchmarks count and simplify the code.

File: runtime-modules/proposals/codex/src/lib.rs [hunk 6]
```diff
@@ -372,31 +538,87 @@ decl_module! {
...
+                    ensure!(amount<=current_budget, Error::<T>::InsufficientFundsForFundingRequest);
+
+                    call_wg!(working_group<T>, set_budget, origin.clone(), wg_budget.saturating_add(amount))?;
+                    Council::<T>::set_budget(origin, current_budget - amount)?;
+                },
+                BalanceKind::Negative => {
+                    ensure!(amount <= wg_budget, Error::<T>::InsufficientFundsForFundingRequest);
+
+                    call_wg!(working_group<T>, set_budget, origin.clone(), wg_budget - amount)?;
+                    Council::<T>::set_budget(origin, current_budget.saturating_add(amount))?;
```
Comment (shamil-gadelshin) [inline runtime-modules/proposals/codex/src/lib.rs:543]: Why does it set the council budget rather than working group budget?

File: runtime-modules/proposals/codex/src/lib.rs [hunk 7]
```diff
@@ -372,31 +538,87 @@ decl_module! {
...
+                }
+            }
+        }
+
+        /// Funding request proposal
+        #[weight = WeightInfoCodex::<T>::funding_request()] // TODO: adjust weight
+        pub fn funding_request(
```
Comment (shamil-gadelshin) [inline runtime-modules/proposals/codex/src/lib.rs]: You don't need this extrinsic. Because you could just run the extrinsic from the council module.

File: runtime-modules/proposals/codex/src/lib.rs [hunk 8]
```diff
@@ -372,31 +538,87 @@ decl_module! {
...
+            ProposalDetails::FundingRequest(ref amount, _) => {
+                ensure!(
+                    *amount != BalanceOf::<T>::zero(),
+                    Error::<T>::InvalidFundingRequestProposalBalance
+                );
+
                 ensure!(
-                    *balance != BalanceOf::<T>::zero(),
-                    Error::<T>::InvalidSpendingProposalBalance
+                    *amount <= <BalanceOf<T>>::from(MAX_SPENDING_PROPOSAL_VALUE),
```
Comment (shamil-gadelshin) [inline runtime-modules/proposals/codex/src/lib.rs:568]: We don't check for the storage of other modules when create the proposal. We check only proposal parameters. Ask Bedeho for details.

**Review by shamil-gadelshin — CHANGES_REQUESTED** [review]
Great work, but there are major issues:
- working group dependency on its own instances
- codex dependency on the working group instances
- unnecessary extrinsic proxies

Also, please, update the codex pallet comment.

File: runtime-modules/proposals/codex/src/types.rs (hunk 4 above)
Comment (shamil-gadelshin) [reply runtime-modules/proposals/codex/src/types.rs:142]: I think that the gitbook should be updated rather than changing this parameter.

File: runtime-modules/proposals/codex/src/lib.rs (hunk 1 above)
Comment (conectado) [reply runtime-modules/proposals/codex/src/lib.rs]: I could fix this by adding a hook to the runtime that allows us to query and set the working group budget. Also, this would fix having the working group instances definitions in the working group since this way I would only need them in the runtime and I can move them back there. Do you think this is an acceptable way of handling this?

File: runtime-modules/proposals/codex/src/benchmarking.rs (hunk 5 above)
Comment (conectado) [reply runtime-modules/proposals/codex/src/benchmarking.rs:506]: We could do that, but I am a little reticent to do it since the parameter would imply an order from best to worst-case scenario and the benchmarks will try to do a linear regression over it and might create and spurious linear dependency.

File: runtime-modules/proposals/codex/src/lib.rs (hunk 6 above)
Comment (conectado) [reply runtime-modules/proposals/codex/src/lib.rs:543]: Since the balance is non-negative in the line before this I reduce the working group budget by `amount` and in this line I add that `amount` to the council budget. That's the behavior I understand is expected [from the handbook](https://joystream.gitbook.io/joystream-handbook/governance/proposals#update-working-group-budget).

File: runtime-modules/proposals/codex/src/types.rs (hunk 4 above)
Comment (conectado) [reply runtime-modules/proposals/codex/src/types.rs:142]: I think this parameter is missing in the gitbook because the generic description for the proposal already doubles as a rationale. I remember bedeho saying he dropped the rationale from slashing lead for this reason.

File: runtime-modules/proposals/codex/src/lib.rs (hunk 8 above)
Comment (conectado) [reply runtime-modules/proposals/codex/src/lib.rs:568]: I see, in the [handbook](https://joystream.gitbook.io/joystream-handbook/governance/proposals#creation-conditions-3) it says that it should check for council balance, so I will ask for clarification

File: runtime-modules/proposals/codex/src/lib.rs (hunk 1 above)
Comment (shamil-gadelshin) [reply runtime-modules/proposals/codex/src/lib.rs]: Yes, I also thought of the runtime hook. Let's create a prototype to see how it goes.

File: runtime-modules/proposals/codex/src/types.rs (hunk 4 above)
Comment (shamil-gadelshin) [reply runtime-modules/proposals/codex/src/types.rs:142]: Ok. Make a comment please when you pass the empty text on two occasions: slashing and termination.

File: runtime-modules/proposals/codex/src/benchmarking.rs (hunk 5 above)
Comment (shamil-gadelshin) [reply runtime-modules/proposals/codex/src/benchmarking.rs:506]: Fair enough. Let it be.

File: runtime-modules/proposals/codex/src/lib.rs (hunk 6 above)
Comment (shamil-gadelshin) [reply runtime-modules/proposals/codex/src/lib.rs:543]: Sorry, I missed the call_wg part.

File: runtime/src/proposals_configuration/tests.rs [hunk 9]
```diff
@@ -1,10 +1,371 @@
 use crate::ProposalParameters;
 
 // Enable during the conditional compilation tests.
+#[test]
+#[ignore]
+fn proposal_parameters_are_initialized_max_validators() {
+    let actual_params = super::SetMaxValidatorCountProposalParameters::get();
+    let expected_params = ProposalParameters {
```
Comment (shamil-gadelshin) [inline runtime/src/proposals_configuration/tests.rs]: Please, consider creating "default_proposal_parameters" function.

File: runtime-modules/working-group/src/lib.rs [hunk 10]
```diff
@@ -1029,6 +1029,16 @@ decl_module! {
 }
 
 impl<T: Trait<I>, I: Instance> Module<T, I> {
+    // Set working group budget
+    pub fn put_budget(new_budget: BalanceOf<T>) {
```
Comment (shamil-gadelshin) [inline runtime-modules/working-group/src/lib.rs]: Please, remove these duplicated methods.

**Review by shamil-gadelshin — CHANGES_REQUESTED** [review]
There are still a couple of minor issues.

**Review by shamil-gadelshin — APPROVED** [review]
LGTM
